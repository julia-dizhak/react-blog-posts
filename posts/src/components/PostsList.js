import React, { Component } from 'react';
import PropTypes from 'prop-types';
//import FlipMove from 'react-flip-move';

import { Button } from '../shared/Button';
import { SORTS } from './../constants/SORTS';
import SortButton from '../shared/ButtonSort';

const largeColumn = {width: '30%'},
    midColumn = {width: '26%'},
    smallColumn = {width: '29%'};

const propTypes = {
    list: PropTypes.arrayOf(
        PropTypes.shape({
            objectID: PropTypes.string.isRequired,
            author: PropTypes.string,
            url: PropTypes.string,
            num_comments: PropTypes.number,
            points: PropTypes.number
        })
    ).isRequired,
    onDismiss: PropTypes.func.isRequired
};

export default class PostsList extends Component {  
    constructor(props) {
        super(props);
        
        this.state = {
            sortKey: 'NONE',
            isSortReverse: false
        };

        this.handleSort = this.handleSort.bind(this);
    }

    handleSort(sortKey) {
        const isSortReverse = this.state.sortKey === sortKey && !this.state.isSortReverse;
        this.setState({sortKey, isSortReverse});
    }

    render() {  
        const {
            list, 
            onDismiss,
            onArchive  
        } = this.props; 
        
        const {
            sortKey, 
            isSortReverse
        } = this.state;

        const sortedList = SORTS[sortKey](list),
            reverseSortedList = isSortReverse ? sortedList.reverse() : sortedList; 

        if ( !list ) {
            return null;
        }

        return (
            <React.Fragment>
                <ul className="posts-list">
                    {reverseSortedList.map(item =>
                        <li 
                            className="posts" 
                            key={item.objectID}>
                            <p className="title">
                                <SortButton
                                    sortKey={'TITLE'}
                                    activeSortKey={sortKey}
                                    isSortReverse={isSortReverse}
                                    handleSort={this.handleSort} 
                                />  
                                <a href={item.url} target="_blank" className="title-link">{item.title}&nbsp;</a>
                                <span className="points">
                                    <span className="points-wrap">
                                        <SortButton
                                            sortKey={'POINTS'}
                                            activeSortKey={sortKey}
                                            isSortReverse={isSortReverse}
                                            handleSort={this.handleSort} 
                                        />
                                        points&nbsp;
                                    </span>
                                    {item.points}
                                </span>
                            </p>

                            <span style={midColumn}>
                                <span>{item.author}&nbsp;</span>
                            </span>
                            
                            <span style={smallColumn}>
                                <SortButton
                                    sortKey={'COMMENTS'}
                                    activeSortKey={sortKey}
                                    isSortReverse={isSortReverse}
                                    handleSort={this.handleSort} 
                                />
                                <span>comments: {item.num_comments}&nbsp;</span>
                            </span>

                            <span style={largeColumn}>
                                <Button
                                    className="button-cancel"
                                    onClick={() => onDismiss(item.objectID)}>
                                    Dismiss
                                </Button>

                                <Button
                                    onClick={() => onArchive(item.objectID)}>
                                    Archive
                                </Button>
                            </span>
                        </li> )   
                    }
              
                </ul>
            </React.Fragment> 
        );
    }    
}

PostsList.propTypes = propTypes;
