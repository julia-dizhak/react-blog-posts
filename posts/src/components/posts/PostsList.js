import React, { Component } from 'react';

import Button from './Button';
import byArchived from './../../utils/archived';

const largeColumn = { width: '40%'},
    midColumn = { width: '35%'},
    smallColumn = { width: '10%'};

// function isSearched(query='') {
//     return function(item) {
//       // filter return new array
//       return !query || item.title.toLowerCase().includes(query.toLowerCase());
//       //return !query || item.title.toLowerCase().indexOf(query.toLowerCase()) !== -1;
//     }
// }

export default class PostsList extends Component {  
    constructor(props) {
        super(props);
        
        this.state = {
          archivedItems: []
        };

        this.onArchive = this.onArchive.bind(this);
    }

    onArchive(id) {
        const { archivedItems } = this.state;
    
        this.setState({
            archivedItems: [...archivedItems, id]
        });
        console.log(archivedItems);
    }

    render() {  
        const { list, onDismiss } = this.props; 
        const { archivedItems } = this.state;

        const filteredList = list.filter(byArchived(archivedItems));
        console.log(filteredList)

        return (
            <React.Fragment>
                <ul className="posts-list">
                {filteredList.map(item =>
                    <li 
                        className="posts" 
                        key={item.objectID}>
                        <span style={largeColumn}>
                            <a href={item.url}>{ item.title }&nbsp;</a>
                        </span>
                        <span style={midColumn}>{ item.author }&nbsp;</span>
                        <span style={smallColumn}>{ item.num_comments }&nbsp;</span>
                        <span style={{width: '10%'}}>{ item.points }&nbsp;</span>
                        <span style={{width: '30%'}}>
                            <Button
                                onClick={() => onDismiss(item.objectID)}>
                                Dismiss
                            </Button>
                            <Button
                                onClick={() => this.onArchive(item.objectID)}>
                                Archive
                            </Button>
                        </span>
                    </li> )   
                }
                </ul> 
            </React.Fragment> 
        );
    }    
}