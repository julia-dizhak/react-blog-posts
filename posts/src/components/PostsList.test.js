import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer'; 
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import PostsList from './PostsList';

Enzyme.configure({ adapter: new Adapter() });

describe('PostsList', () => {
    const props = {
        list: [
            { title: '1', author: '1', num_comments: 1, points: 2, objectID: 'y' },
            { title: '2', author: '2', num_comments: 2, points: 3, objectID: 'z' },
        ],
        onDismiss: () => {},
        sortKey: 'TITLE',
        isSortReverse: false
    };

    it('renders without crashing', () => {
        const div = document.createElement('div');
        ReactDOM.render(<PostsList {...props} />, div);
    });
    
    test('has a valid snapshot', () => {
        const component = renderer.create(
            <PostsList {...props} />
        );
        
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('shows two items in list', () => {
        const element = shallow(
            <PostsList {...props} />
        );

        expect(element.find('.posts-list li').length).toBe(2);
    });
});
