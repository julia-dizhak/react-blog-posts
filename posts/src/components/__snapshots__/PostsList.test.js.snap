// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostsList has a valid snapshot 1`] = `
<ul
  className="posts-list"
>
  <li
    className="posts"
  >
    <span
      className="title"
      style={
        Object {
          "width": "30%",
        }
      }
    >
      <a
        href={undefined}
      >
        1
         
      </a>
    </span>
    <span
      style={
        Object {
          "width": "15%",
        }
      }
    >
      1
       
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      1
       
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      2
       
    </span>
    <span
      style={
        Object {
          "width": "30%",
        }
      }
    >
      <button
        className="button-cancel"
        onClick={[Function]}
      >
        Dismiss
      </button>
      <button
        className="button"
        onClick={[Function]}
      >
        Archive
      </button>
    </span>
  </li>
  <li
    className="posts"
  >
    <span
      className="title"
      style={
        Object {
          "width": "30%",
        }
      }
    >
      <a
        href={undefined}
      >
        2
         
      </a>
    </span>
    <span
      style={
        Object {
          "width": "15%",
        }
      }
    >
      2
       
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      2
       
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      3
       
    </span>
    <span
      style={
        Object {
          "width": "30%",
        }
      }
    >
      <button
        className="button-cancel"
        onClick={[Function]}
      >
        Dismiss
      </button>
      <button
        className="button"
        onClick={[Function]}
      >
        Archive
      </button>
    </span>
  </li>
</ul>
`;
